File: src/lib/cr-compliance-config.ts
// Project configuration for CR Compliance — unlike CP10/CP38 (one fixed
// Jira ticket per project), this checkpoint answers ad hoc questions about
// *any* change-request ticket, so the ticket key comes from the question
// text and only the GitHub repo it maps to is fixed config.
export type CrComplianceProjectConfig = {
  projectId: string;
  projectLabel: string;
  checkpointId: "CR";
  // Jira project key (e.g. "POC") that this GitHub repo's tickets live under.
  jiraProjectKey: string;
  githubOwner: string;
  githubRepo: string;
  // id of the matching entry in the checkpoint ledger (audit-data.ts) —
  // lets a definitive result update that checkpoint's status, same as
  // CP10/CP38.
  ledgerId: string;
};

export const crComplianceConfigs: CrComplianceProjectConfig[] = [
  {
    projectId: "project-a",
    projectLabel: "PIT Armour",
    checkpointId: "CR",
    jiraProjectKey: "POC",
    githubOwner: "Anushka34567",
    githubRepo: "Github_integration",
    // cp-18 in the ledger ("Git commits are traceable to approved CRs") is
    // exactly this control — reused rather than adding a duplicate line
    // item for the same requirement.
    ledgerId: "cp-18",
  },
];

// Jira issue keys are PROJECTKEY-NUMBER. Users naturally write "CR-POC-4" or
// "CR POC-4" to mean "change request POC-4" — the CR- is describing the
// ticket type, not part of the real key — so it's stripped when present,
// but a bare "POC-4" resolves the same way.
const TICKET_KEY_RE = /\bcr[-\s]?([a-z][a-z0-9]{1,9}-\d+)\b|\b([a-z][a-z0-9]{1,9}-\d+)\b/i;

export function extractTicketKey(question: string): string | null {
  const m = question.match(TICKET_KEY_RE);
  const key = m?.[1] ?? m?.[2];
  return key ? key.toUpperCase() : null;
}

// Deterministic routing, same approach as CP10/CP38: never LLM-guessed.
export function looksLikeCrComplianceQuestion(question: string): boolean {
  const q = question.toLowerCase();
  const mentionsCompliance = q.includes("compliant") || q.includes("compliance");
  const mentionsTicket = q.includes("ticket") || q.includes("cr-") || q.includes("cr ");
  return mentionsCompliance && mentionsTicket && extractTicketKey(question) !== null;
}

export function resolveCrComplianceProject(ticketKey: string): CrComplianceProjectConfig | null {
  const prefix = ticketKey.split("-")[0]?.toUpperCase();
  return crComplianceConfigs.find((c) => c.jiraProjectKey.toUpperCase() === prefix) ?? null;
}

File: backend/app/config/cr_compliance_config.py
# Project configuration for CR Compliance — unlike CP10/CP38 (one fixed
# Jira ticket per project), this checkpoint answers ad hoc questions about
# *any* change-request ticket, so the ticket key comes from the question
# text and only the GitHub repo it maps to is fixed config.
import re

from app.models import CrComplianceProjectConfig

cr_compliance_configs: list[CrComplianceProjectConfig] = [
    CrComplianceProjectConfig(
        projectId="project-a",
        projectLabel="PIT Armour",
        checkpointId="CR",
        jiraProjectKey="POC",
        githubOwner="Anushka34567",
        githubRepo="Github_integration",
        # cp-18 in the ledger ("Git commits are traceable to approved CRs") is
        # exactly this control — reused rather than adding a duplicate line
        # item for the same requirement.
        ledgerId="cp-18",
    ),
]

# Jira issue keys are PROJECTKEY-NUMBER. Users naturally write "CR-POC-4" or
# "CR POC-4" to mean "change request POC-4" — the CR- is describing the
# ticket type, not part of the real key — so it's stripped when present,
# but a bare "POC-4" resolves the same way.
_TICKET_KEY_RE = re.compile(r"\bcr[-\s]?([a-z][a-z0-9]{1,9}-\d+)\b|\b([a-z][a-z0-9]{1,9}-\d+)\b", re.IGNORECASE)


def extract_ticket_key(question: str) -> str | None:
    m = _TICKET_KEY_RE.search(question)
    if not m:
        return None
    key = m.group(1) or m.group(2)
    return key.upper() if key else None


def resolve_cr_compliance_project(ticket_key: str) -> CrComplianceProjectConfig | None:
    prefix = ticket_key.split("-")[0].upper() if ticket_key else None
    for cfg in cr_compliance_configs:
        if cfg.jira_project_key.upper() == prefix:
            return cfg
    return None
